tensorboard/plugins/debugger/tensor_store.py

```diff
@@ -0,0 +1,331 @@
+        # Always keep at least one time index in the memory.
+        break
+      i -= 1
+    # i is now the last time index to defer to disk.
+
+    j = i
+    cum_disk_size = 0
+    while j >= 0:
+      cum_disk_size += self._data[j].nbytes
+      if j < i and cum_disk_size > self._disk_bytes_limit:
+        break
+      j -= 1
+    # j is now the last time index to discard (due to limit).
+
+    # Defer some tensors to disk.
+    while i > j:
+      if not isinstance(self._data[i], _TensorValueDeferredToDisk):
+        deferred = _TensorValueDeferredToDisk(
+            self._watch_key, i, self._data[i], self._root_dir)
+        self._data[i] = deferred
+      i -= 1
+
+    # Mark remaining ones as discarded.
+    while i >= 0:
+      if not isinstance(self._data[i], _TensorValueDiscarded):
+        if isinstance(self._data[i], _TensorValueDeferredToDisk):
+          self._data[i].dispose()
+        self._data[i] = _TensorValueDiscarded(self._watch_key, i)
+      i -= 1
+
+  def num_total(self):
+    """Get the total number of values."""
+    return len(self._data)
+
+  def num_in_memory(self):
+    """Get number of values in memory."""
+    n = len(self._data) - 1
+    while n >= 0:
+      if isinstance(
+          self._data[n],
+          (_TensorValueDeferredToDisk, _TensorValueDiscarded)):
+        break
+      n -= 1
+    return len(self._data) - 1 - n
+
+  def num_on_disk(self):
+    """Get the number of values currently stored on disk."""
+    return len(self._data) - self.num_in_memory() - self.num_discarded()
+
+  def num_discarded(self):
+    """Get the number of values discarded due to exceeding both limits."""
+    if not self._data:
+      return 0
+    n = 0
+    while n < len(self._data):
+      if not isinstance(self._data[n], _TensorValueDiscarded):
+        break
+      n += 1
+    return n
+
+  def query(self, time_indices):
+    """Query the values at given time indices.
+
+    Args:
+      time_indices: 0-based time indices to query, as a `list` of `int`.
+
+    Returns:
+      Values as a list of `numpy.ndarray` (for time indices in memory or on
+      disk) or `None` (for time indices discarded).
+    """
+    if self._disposed:
+      raise ValueError(
+          'Cannot query: this _WatchStore instance is already disposed')
+    if not isinstance(time_indices, (tuple, list)):
+      time_indices = [time_indices]
+    output = []
+    for time_index in time_indices:
+      if isinstance(self._data[time_index], _TensorValueDiscarded):
+        output.append(None)
+      elif isinstance(self._data[time_index], _TensorValueDeferredToDisk):
+        output.append(self._data[time_index].get_value())
+      else:
+        output.append(self._data[time_index])
+    return output
+
+  def dispose(self):
+    for datum in self._data:
+      if isinstance(datum, _TensorValueDeferredToDisk):
+        datum.dispose()
+    self._disposed = True
+
+
+class TensorStore(object):
```

Comment:
Can we punt on checking this file in? And simply have it store in memory with an upper bound on usage? After which it discards.